python/cudf/cudf/core/index.py

```diff
@@ -2744,6 +2749,216 @@ def categories(self):
+        if end:
+            start = end - (freq * periods)
+        if start:
+            end = freq * periods + start
+        left_col = arange(start, end, freq)
+        assert end is not None and start is not None
+        end = end + 1
+        start = start + freq
+        right_col = arange(start, end, freq)
+    elif freq and not periods:
+        assert end is not None and start is not None
+        end = end - freq + 1
+        left_col = arange(start, end, freq)
+        end = end + freq + 1
+        start = start + freq
+        right_col = arange(start, end, freq)
+    elif start is not None and end is not None:
+        left_col = arange(start, end, freq)
+        start = start + 1
+        end = end + 1
+        right_col = arange(start, end, freq)
+    else:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and "
+            "freq, atleast two must be specified"
+        )
+    if len(right_col) == 0 or len(left_col) == 0:
+        return cudf.IntervalIndex([], closed=closed)
+
+    interval_col = column.build_interval_column(
+        left_col, right_col, closed=closed
+    )
+    return IntervalIndex(interval_col)
+
+
+class IntervalIndex(GenericIndex):
+    """
+    Immutable index of intervals that are closed on the same side.
+
+    Parameters
+    ----------
+    data : array-like (1-dimensional)
+        Array-like containing Interval objects from which to build the
+        IntervalIndex.
+
+    closed : {"left", "right", "both", "neither"}, default "right"
+        Whether the intervals are closed on the left-side, right-side,
+        both or neither.
+
+    dtype : dtype or None, default None
+        If None, dtype will be inferred.
+
+    copy : bool, default False
+        Copy the input data.
+
+    name : object, optional
+        Name to be stored in the index.
+
+    Returns
+    -------
+    IntervalIndex
+    """
+
+    def __new__(
+        cls, data=None, closed=None, dtype=None, copy=False, name=None,
+    ) -> "IntervalIndex":
+        if copy:
+            data = column.as_column(data, dtype=dtype).copy()
+        out = Frame.__new__(cls)
+        kwargs = _setdefault_name(data, name=name)
+        if isinstance(data, IntervalColumn):
+            data = data
+        elif isinstance(data, pd.Series) and (is_interval_dtype(data.dtype)):
+            data = column.as_column(data, data.dtype)
+        elif isinstance(data, (pd._libs.interval.Interval, pd.IntervalIndex)):
+            data = column.as_column(data, dtype=dtype,)
+        elif data is not None and data != [] and data[0].closed != closed:
+                # when closed is not the same as the data's closed
+                # we need to change the data closed value,
+                left_col = [data[i].left for i in range(len(data))]
+                # creating a col out of the left child so we can get
+                # the correct dtype
+                left = column.as_column(left_col)
```

Comment:
This is really expensive to be able to try to infer the dtype of the `data` input object. We shouldn't need to copy to the GPU in order to do this.

I'd expect our `column.as_column` function to handle this dtype inference for us instead of us having to do a separate `column.as_column` for this `data`.

Reply:
Yes, that's fair. I'll work on an alternative to this!